crates/artifacts/resolc/src/contract.rs

```diff
@@ -0,0 +1,92 @@
+use std::collections::{BTreeMap, HashSet};
+
+use alloy_json_abi::JsonAbi;
+use foundry_compilers_artifacts_solc::{DevDoc, LosslessMetadata, StorageLayout, UserDoc};
+use serde::{Deserialize, Serialize};
+
+use crate::ResolcEVM;
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct ResolcContract {
+    /// The contract ABI.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub abi: Option<JsonAbi>,
+    /// The contract metadata.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub metadata: Option<serde_json::Value>,
+    /// The contract developer documentation.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub devdoc: Option<DevDoc>,
+    /// The contract user documentation.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub userdoc: Option<UserDoc>,
+    /// The contract storage layout.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub storage_layout: Option<StorageLayout>,
+    /// Contract's bytecode and related objects
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub evm: Option<ResolcEVM>,
+    /// The contract optimized IR code.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub ir_optimized: Option<String>,
+    /// The contract PolkaVM bytecode hash.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub hash: Option<String>,
+    /// The contract factory dependencies.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub factory_dependencies: Option<BTreeMap<String, String>>,
+    /// The contract missing libraries.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub missing_libraries: Option<HashSet<String>>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Default)]
+struct Output {
+    devdoc: DevDoc,
+    userdoc: UserDoc,
+    abi: JsonAbi,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Default)]
+struct Meta {
+    output: Output,
+}
+
+impl From<ResolcContract> for foundry_compilers_artifacts_solc::Contract {
```

Comment:
I would say instead of using the from we implement custom deserializer to ensure we check for direct abi/userdoc/devdoc and and fields before falling back to metadata values since some revive outputs include these directly.

Reply:
Can you provide a snippet of solidity code where upon compiling we actually get the docs back on toplevel struct?

Reply:
was done as a fix for output selection on revive side